File: 2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        List<Integer> cpoints = new ArrayList<>();
        ListNode prev = head;
        ListNode curr = head.next;
        int idx = 1;

        while (curr != null && curr.next != null) {
            if ((prev.val < curr.val && curr.val > curr.next.val) || (prev.val > curr.val && curr.val < curr.next.val)) {
                cpoints.add(idx);
            }
            prev = curr;
            curr = curr.next;
            idx++;
        }

        if (cpoints.size() < 2) {
            return new int[]{-1, -1};
        }

        int minDist = Integer.MAX_VALUE;
        for (int i = 1; i < cpoints.size(); i++) {
            int dist = cpoints.get(i) - cpoints.get(i - 1);
            minDist = Math.min(minDist, dist);
        }

        int maxDist = cpoints.get(cpoints.size() - 1) - cpoints.get(0);

        return new int[]{minDist, maxDist};
    }
}

File: 2182-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def nodesBetweenCriticalPoints(self, head):
        cpoints = []
        prev = head
        curr = head.next
        idx = 1

        while curr and curr.next:
            if (prev.val < curr.val > curr.next.val) or (prev.val > curr.val < curr.next.val):
                cpoints.append(idx)
            prev = curr
            curr = curr.next
            idx += 1
        
        if len(cpoints) < 2:
            return [-1, -1]
        
        min_dist = float('inf')

        for i in range(1, len(cpoints)):
            dist = cpoints[i] - cpoints[i - 1]
            min_dist = min(min_dist, dist)
        
        max_dist = cpoints[-1] - cpoints[0]
        
        return [min_dist, max_dist]
